src/lib/tls/tls13/tls_client_impl_13.cpp

```diff
@@ -421,18 +476,34 @@ void Client_Impl_13::handle(const Finished_13& finished_msg)
    m_cipher_state->advance_with_server_finished(m_transcript_hash.previous());
    m_cipher_state->advance_with_client_finished(m_transcript_hash.current());
 
-   // TODO: save session and invoke tls_session_established callback
+   // TODO: Create a dummy session object and invoke tls_session_established.
+   //       Alternatively, consider changing the expectations described in the
+   //       callback's doc string.
```

Comment:
The callback `tls_session_established` has the underlying assumption that a successful TLS handshake generates exactly one resumable session. In TLS 1.2 this session was constructed and reported via this callback. TLS 1.3 breaks with this assumption by allowing an arbitrary number of "New Session Ticket" messages at the server's discretion.

Therefore, I feel this callback doesn't really fit for TLS 1.3. Creating a "dummy" session as suggested by the comment might be a way out, but it feels wrong...

Reply:
Repeated "New Session Ticket" messages establishe multiple resumable session tickets. I.e. each session ticket can be used to resume a TLS session independently. The cryptographic parameters are fixed after the handshake and won't differ between individual session tickets. However: The server might opt to hand out additional session tickets after certain events have occured (e.g. after a post-handshake client authentication).

To the best of my knowledge, this behaviour is application-specific, though. RFC 8446 just specifies the possibility to hand out an arbitrary number of session tickets at any point after the handshake.